internal/statesync/statesync.go

```diff
@@ -0,0 +1,316 @@
+)
+
+// StateSyncer is responsible for initializing the database state from the
+// snapshots received from the network. It validates the snapshots against
+// the trusted snapshot providers.
+// The snapshot used to initialize are discarded after the database is restored.
+// Snapshot store if enabled, only stores snapshots produced by the node itself.
+
+type StateSyncer struct {
+	// statesync configuration
+	dbConfig *DBConfig
+
+	// directory to store snapshots and chunks
+	// same as the snapshot store directory
+	// as we allow to reuse the received snapshots
+	snapshotsDir string
+
+	// trusted snapshot providers for verification - cometbft rfc servers
+	snapshotProviders []*rpchttp.HTTP
+
+	// State syncer state
+	inProgress bool
+	snapshot   *Snapshot
+	chunks     map[uint32]bool // Chunks received till now
+	rcvdChunks uint32          // Number of chunks received till now
+
+	// Logger
+	log log.Logger
+}
+
+// NewStateSyncer will initialize the state syncer that enables the node to
+// receive and validate snapshots from the network and initialize the database state.
+// It takes the database configuration, snapshot directory, and the trusted snapshot providers.
+// Trusted snapshot providers are special nodes in the network trusted by the nodes and
+// have snashopt creation enabled. These nodes are responsible for creating and validating snapshots.
+func NewStateSyncer(ctx context.Context, cfg *DBConfig, snapshotDir string, providers []string, logger log.Logger) *StateSyncer {
+
+	ss := &StateSyncer{
+		dbConfig:     cfg,
+		snapshotsDir: snapshotDir,
+		log:          logger,
+	}
+
+	for _, s := range providers {
+		clt, err := ChainRPCClient(s)
+		if err != nil {
+			logger.Error("Failed to create rpc client", log.String("server", s), log.Error(err))
+			return nil
+		}
+		ss.snapshotProviders = append(ss.snapshotProviders, clt)
+	}
+
+	// Ensure that the snasphot directory exists and is empty
+	if err := os.RemoveAll(snapshotDir); err != nil {
+		logger.Error("Failed to delete snapshot directory", log.String("dir", snapshotDir), log.Error(err))
+		return nil
+	}
+	if err := os.MkdirAll(snapshotDir, 0755); err != nil {
+		logger.Error("Failed to create snapshot directory", log.String("dir", snapshotDir), log.Error(err))
+		return nil
+	}
+
+	return ss
+}
+
+// OfferSnapshot checks if the snapshot is valid and kicks off the state sync process
+// accepted snapshot is stored on disk for later processing
+func (ss *StateSyncer) OfferSnapshot(snapshot *Snapshot) error {
+	ss.log.Info("Offering snapshot", log.Int("height", int64(snapshot.Height)), log.Uint("format", snapshot.Format), log.String("App Hash", fmt.Sprintf("%x", snapshot.SnapshotHash)))
+
+	// Check if we are already in the middle of a snapshot
+	if ss.inProgress {
+		return ErrStateSyncInProgress
+	}
+
+	// Validate the snapshot
+	err := ss.validateSnapshot(*snapshot)
+	if err != nil {
+		return err
+	}
+
+	ss.snapshot = snapshot
+	ss.inProgress = true
```

Comment:
The possible concurrency of calls to `OfferSnapshot` and `ApplySnapshotChunk`, or multiple concurrent calls to either one is unclear to me.

Do we know that one call to `OfferSnapshot` must return before it can be called again?

The concerns are with a number of fields including `inProgress` and `chunks`.

Maybe it's not a concern at all.  Depends on how these methods are used by cometbft.

Reply:
These are not concurrently called at all.